### PR TITLE
Fix `HashId` with inodes

### PR DESCRIPTION
**Pull request by sebastiencs**
- Fix HashId created before commit on inodes.
  They are recomputed with the correct index (index on disk) and by removing the working tree bit
- Do not recompute hashes from deserialized inodes

### CODE REVIEW

File: tezos/context/src/serialize/mod.rs
Comment (tizoc) [inline tezos/context/src/serialize/mod.rs:323]: @sebastiencs how does this work? I am not clear on this logic here with a possible 0 and also the next line that gets the max hash id.

File: tezos/context/src/serialize/persistent.rs
Comment (tizoc) [inline tezos/context/src/serialize/persistent.rs:1697]: `HashId::try_new` would be a better name.

File: tezos/context/src/working_tree/storage.rs
Comment (tizoc) [inline tezos/context/src/working_tree/storage.rs:404]: How does this work?

File: tezos/context/src/serialize/mod.rs
Comment (sebastiencs) [reply tezos/context/src/serialize/mod.rs:323]: A `HashId` serialized to `0` will be deserialized as `Option::<HashId>::None`.
We have to handle this case for the parent of a commit, sometimes it can have no parent:
https://github.com/tezedge/tezedge/blob/develop/tezos/context/src/serialize/in_memory.rs#L538

The max hash id is only for statistics, it was useful for the in-memory context, to get an idea of the highest `HashId`

File: tezos/context/src/serialize/persistent.rs
Comment (sebastiencs) [reply tezos/context/src/serialize/persistent.rs:1697]: hmm not sure about that, `HashId::new` is only a call to `NonZeroU64::new`:
https://github.com/tezedge/tezedge/blob/develop/tezos/context/src/kv_store/mod.rs#L60
`NonZeroU64::new` is not `NonZeroU64::try_new`

File: tezos/context/src/working_tree/storage.rs
Comment (sebastiencs) [reply tezos/context/src/working_tree/storage.rs:404]: This is to make sure that when we call, `PointerToInodeInner::get_offset`, the method return a valid (previously set) offset.
Without this bit, `get_offset` could return an invalid offset and we would have no error.
https://github.com/tezedge/tezedge/pull/889/files#diff-cfe31f13c183c59ed45a323b4b25dffd667af5d4b4d5d9e5af3cbd0a30a965b0R505

File: tezos/context/src/serialize/persistent.rs
Comment (tizoc) [reply tezos/context/src/serialize/persistent.rs:1697]: You are right. I always assumed `try_new` was expected when failure was possible.

File: tezos/context/src/working_tree/storage.rs
Comment (tizoc) [reply tezos/context/src/working_tree/storage.rs:404]: I see, makes sense. But wouldn't it be better to use some sentinel value (like `MAX(u64)`) and not lose the bit? or what about 0? it isn't a valid offset right? (these are negative right? not absolute positions in the file)

File: tezos/context/src/working_tree/storage.rs
Comment (tizoc) [inline tezos/context/src/working_tree/storage.rs:439]: @sebastiencs I see here for example that 0 is used when the offset is not available, cannot that work as a sentinel value?

File: tezos/context/src/working_tree/storage.rs
Comment (sebastiencs) [reply tezos/context/src/working_tree/storage.rs:404]: Ah yes, you are right, I will use `0` as sentinel value.

> these are negative right? not absolute positions in the file

Note that here, in `PointerToInodeInner`, the offset is the absolute offset, not relative.

Relative offset are used in the file, but when they are deserialized into the working tree (`PointerToInodeInner` and `DirEntryInner`), they are directly converted to absolute offset

Comment (tizoc) [reply tezos/context/src/working_tree/storage.rs:404]: > Relative offset are used in the file, but when they are deserialized into the working tree (PointerToInodeInner and DirEntryInner), they are directly converted to absolute offset

Ok, related to this, remember that we don't want to load inodes that are not going to be used unnecessarily, in H that becomes a big issue is not handled properly.

**Comment by tizoc** [comment]
@sebastiencs I don't have anything else to comment for now. Once ready lets merge it, ideally before the release so that it is included (this one is more or a priority than the integrity check IMO)

**Review by tizoc — APPROVED** [review]

**Comment by sebastiencs** [comment]
@tizoc I will merge once the CI on Drone pass

**Comment by tizoc** [comment]
@sebastiencs will need to be rebased before merging

**Comment by tizoc** [comment]
Other than that every relevant test on drone passed

**Pull request merged**
